File: src/mtgsim/api/models/deck.py
"""Deck-related Pydantic models."""

from pydantic import BaseModel, Field

from mtgsim.api.models.common import (
    HistogramBucket,
    KeywordCounts,
    Pagination,
)


class DeckLegality(BaseModel):
    """Format legality for a deck."""

    standard: bool = False
    pioneer: bool = False
    modern: bool = False
    legacy: bool = False
    vintage: bool = False
    commander: bool = False
    brawl: bool = False
    historic: bool = False
    pauper: bool = False


class DeckSummary(BaseModel):
    """Summary deck information for list views."""

    file: str
    name: str
    code: str
    card_count: int
    colors: list[str]
    price: float | None = None
    release_date: str | None = None
    legality: DeckLegality


class DeckFilters(BaseModel):
    """Available filter options for deck list."""

    formats: list[str]
    sets: list[str]
    color_combinations: list[list[str]]


class DeckListResponse(BaseModel):
    """Response for deck list endpoint."""

    data: list[DeckSummary]
    pagination: Pagination
    filters: DeckFilters


class DeckCard(BaseModel):
    """Card within a deck."""

    uuid: str
    name: str
    count: int = 1
    mana_cost: str | None = None
    mana_value: int = 0
    type: str | None = None
    rarity: str | None = None
    text: str | None = None
    price: float | None = None
    image_url: str | None = None


class PriceBySource(BaseModel):
    """Price breakdown by source."""

    tcgplayer: float | None = None
    cardkingdom: float | None = None
    cardsphere: float | None = None
    cardmarket: float | None = None
    mtgo: float | None = None


class DeckPrice(BaseModel):
    """Deck price information."""

    total: float
    by_source: PriceBySource


class ManaCurve(BaseModel):
    """Mana curve distribution."""

    zero: int = Field(0, alias="0")
    one: int = Field(0, alias="1")
    two: int = Field(0, alias="2")
    three: int = Field(0, alias="3")
    four: int = Field(0, alias="4")
    five: int = Field(0, alias="5")
    six_plus: int = Field(0, alias="6+")

    model_config = {"populate_by_name": True}


class DeckStats(BaseModel):
    """Comprehensive deck statistics."""

    total_cards: int
    unique_cards: int
    mana_curve: dict[str, int]
    type_distribution: dict[str, int]
    rarity_distribution: dict[str, int]
    color_distribution: dict[str, int]
    price_histogram: list[HistogramBucket]
    keywords: KeywordCounts


class DeckMeta(BaseModel):
    """Deck metadata."""

    file: str
    name: str
    code: str
    release_date: str | None = None


class DeckDetail(BaseModel):
    """Full deck details."""

    meta: DeckMeta
    legality: DeckLegality
    colors: list[str]
    price: DeckPrice
    commander: list[DeckCard]
    main_board: list[DeckCard]
    side_board: list[DeckCard]
    stats: DeckStats

File: frontend/src/types/deck.ts
import type { HistogramBucket, KeywordCounts, Pagination } from "./common"

export interface DeckSummary {
  file: string
  name: string
  code: string | null
  deck_type: string | null
  card_count: number
  colors: string[]
  price: number | null
  release_date: string | null
  legality: Record<string, boolean>
  source: string
}

export interface DeckListResponse {
  data: DeckSummary[]
  pagination: Pagination
  filters: {
    sets: string[]
    formats: string[]
    color_combinations: string[]
  }
}

export interface DeckCard {
  uuid: string
  name: string
  count: number
  board: string
  mana_cost: string | null
  mana_value: number
  type: string
  types: string[]
  colors: string[]
  rarity: string
  tags: string[]
  text: string | null
  price: number | null
  image_url: string | null
  owns_enough: boolean
  owned_count: number
  missing_count: number
}

export interface DeckMeta {
  name: string
  file: string
  code: string | null
  release_date: string | null
  description: string | null
  format: string | null
  source: string
}

export interface DeckLegality {
  standard: boolean
  pioneer: boolean
  modern: boolean
  legacy: boolean
  vintage: boolean
  commander: boolean
  brawl: boolean
  historic: boolean
  pauper: boolean
}

export interface DeckPrice {
  total: number
  tcgplayer: number | null
  cardkingdom: number | null
  cardsphere: number | null
  cardmarket: number | null
  mtgo: number | null
}

export interface DeckStats {
  total_cards: number
  unique_cards: number
  mana_curve: Record<string, number>
  type_distribution: Record<string, number>
  rarity_distribution: Record<string, number>
  color_distribution: Record<string, number>
  price_histogram: HistogramBucket[]
  keywords: KeywordCounts
}

export interface DeckDetail {
  meta: DeckMeta
  legality: DeckLegality
  colors: string[]
  price: DeckPrice
  commander: DeckCard[]
  main_board: DeckCard[]
  side_board: DeckCard[]
  stats: DeckStats
}

export interface UserDeckResponse {
  id: number
  name: string
  description: string | null
  format: string | null
  source: string
  card_count: number
}

export interface DeckImportResult {
  deck_id: string
  deck_name: string
  total_cards: number
  resolved: {
    name: string
    matched_name: string
    match_type: "exact" | "fuzzy" | "created"
    match_score: number
    count: number
  }[]
  legality: {
    format: string
    legal: boolean
    reason: string | null
  }[]
}
